### PR TITLE
Poetry: select olddeps using `poetry`

### PR DESCRIPTION
**Pull request by DMRobertson**
There are a few moving parts here.

- Once we ditch setup.py we'll need to patch the dependencies in pyproject.toml to maintain this CI job.
- If we were using poetry-core as our PEP 517 build system, it would suffice to `pip install .[all,test]`. But we're gradually transitioning to that; setuptools remains the build system for now.
- Therefore use the "full" poetry to _lock_ an olddeps environment and run using `poetry run`.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -105,29 +105,7 @@ commands =
...
 usedevelop=true
 
-# A test suite for the oldest supported versions of Python libraries, to catch
-# any uses of APIs not available in them.
-[testenv:py3-old]
-skip_install = true
-usedevelop = false
-deps =
-    Automat == 0.8.0
-    lxml
```
Comment (clokep) [inline tox.ini:115]: I don't see these being moved into the new bits, but maybe that's OK?

Comment (DMRobertson) [reply tox.ini:115]: It should be okay.

- Automat isn't a direct dependency (it's not in python_dependencies,py nor pyproject.toml), so I don't think we have any business pinning it to a particular version anyway;
- lxml will be installed [it's part of `all`](https://github.com/matrix-org/synapse/blob/dd5cc37aa465df8e33fd872de2528dc53740b2a6/pyproject.toml#L217-L227), and in the config we install `all` extras
- markupsafe is also not a direct dependency.
- the base dependencies are pip, setuptools, and stuff related to `coverage`. We've stopped using coverage (and I suspect it'd need some config fiddling to have it work in poetry anyway).

File: .ci/scripts/test_old_deps.sh [hunk 2]
```diff
@@ -9,12 +12,74 @@ set -ex
...
+# The sed expression 's/\^/==/g' ought to do the trick. But it would also change
+# `python = "^3.7"` to `python = "==3.7"`, which would mean we fail because olddeps
+# runs on 3.8 (#12343).
+
+# TODO: I'd prefer to use something like this
+#   https://github.com/python-poetry/poetry/issues/3527
+#   https://github.com/pypa/pip/issues/8085
+# rather than this sed script.
+
+sed -i-backup \
```
Comment (richvdh) [inline .ci/scripts/test_old_deps.sh:43]: is there any point in keeping a backup here?

```suggestion
sed -i \
```

**Review by richvdh — APPROVED** [review]
lgtm!

File: .ci/scripts/test_old_deps.sh (hunk 2 above)
Comment (DMRobertson) [reply .ci/scripts/test_old_deps.sh:41]: Argh, I could have sworn I already removed this somewhere else.

(A previous version of this script diffed the pre- and post patch toml. Unfortunately `toml.dump`'s output style is different to our input, so the resulting diff isn't very meaningful.)

**Review by richvdh — APPROVED** [review]
lgtm otherwise

File: .ci/scripts/test_old_deps.sh [hunk 3]
```diff
@@ -20,42 +20,38 @@ export LANG="C.UTF-8"
...
+#   https://github.com/python-poetry/poetry/issues/3527
+#   https://github.com/pypa/pip/issues/8085
+# to select the lowest possible versions, rather than resorting to this sed script.
+
 # Patch the project definitions in-place:
 # - Replace all lower and tilde bounds with exact bounds
-# - Make the pyopenssl 17.0, which is the oldest version that works with
-#   a `cryptography` compiled against OpenSSL 1.1.
+# - Make the pyopenssl 17.0, which can work against an
+#   OpenSSL 1.1 compiled cryptography (as older ones don't compile on Travis).
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply .ci/scripts/test_old_deps.sh:31]: It is. Thanks.

**Pull request merged**